Step-by-step documentation draft (#2793)
kubeapps/kubeapps: docs/step-by-step/kubeapps-on-tkg/step-2.md

   ```yaml
   clusters:
     - name: my-tkg-cluster
       pinnipedConfig:
         enable: true
   ```

2. Enable the _Pinniped Proxy_ component so that the requests performed by Kubeapps can be proxied through Pinniped, by setting the parameter `pinnipedProxy.enabled=true`. Here is an example:

   ```yaml
   pinnipedProxy:
     enabled: true
     defaultAuthenticatorName: kubeapps-jwt-authenticator # this name must match the authenticator name previously created
     defaultPinnipedNamespace: vmware-system-tmc
     defaultPinnipedAPISuffix: pinniped.tmc.cloud.vmware.com
   ```

   > **TIP**: The `defaultAuthenticatorName` must match the _JWTAuthenticator_ resource name created in [Step 1](./step-1.md).

3. Configure the _OAuth2Proxy_ component by entering the information gathered from the OIDC provider in [Step 1](./step-1.md). This component performs the authentication flow, generating the appropriate request to the login page and retrieving the token in the callback URL. Here is an example. Remember to replace the placeholders as follows:

>    - Replace the `OIDC-ISSUER-URL` with the issuer URL of the OIDC provider. For CSP it is `https://console-stg.cloud.vmware.com/csp/gateway/am/api`.

Reviewer: Staging again...

   - Replace `CLIENT-ID` with the application ID obtained from the JSON file in the previous step.
   - Replace `CLIENT-SECRET` with the application secret obtained from the JSON file in the previous step.
   - Replace `COOKIE-SECRET` with a seed string for secure cookies (should be a 16-, 24-, or 32-byte string).

   ```yaml
   authProxy:
     enabled: true
     provider: oidc
     clientID: CLIENT-ID
     clientSecret: CLIENT-SECRET
     cookieSecret: COOKIE-SECRET
     additionalFlags:
       - --oidc-issuer-url=OIDC-ISSUER-URL
       - --scope=openid email groups
       - --set-authorization-header=true
       # - --insecure-oidc-skip-issuer-verification=true
   ```

>    > **NOTE**: In some providers whose issuer URL does not match the token URL (such as VMware CSP), the flag `--insecure-oidc-skip-issuer-verification=true` must be turned on. Be aware of the security concerns of enabling this flag, which are discussed in the [official OAuth2Proxy documentation](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/).

Reviewer: ```suggestion
   > **NOTE**: In some providers whose issuer URL does not match the token URL (such as VMware CSP), the flag `--insecure-oidc-skip-issuer-verification=true` is currently required. Be aware of the security concerns of enabling this flag, which are discussed in the [official OAuth2Proxy documentation](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/).
```

Author: Have a look at the end of https://github.com/kubeapps/kubeapps/issues/2686#issuecomment-846966217, have you ever tried that manual config instead of relying on the oidc autodiscovery?
If it works, I guess we can send another PR removing all there references to the --insecure flag

> **NOTE**: RBAC configuration depends on your custom business requirements. The configuration shown below is only an example and not meant for production use. Please refer to the official [Kubernetes RBAC documentation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) for more details.

The configuration shown below demonstrates how to create a `ClusterRoleBinding` named `kubeapps-operator` with the `cluster-admin` role for a specified user. Replace the `EMAIL-ADDRESS` placeholder with the email address for the user, as specified in the OIDC provider and name the file `kubeapps-rbac.yaml`.

```yaml
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: kubeapps-operator
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - apiGroup: rbac.authorization.k8s.io
    kind: User
    name: EMAIL-ADDRESS
```

Apply this configuration by executing the following command:

```bash
kubectl apply -f kubeapps-rbac.yaml
```

> At this point, the user having `EMAIL-ADDRESS` email account will have `cluster-admin` access and will be able to perform any desired action via the Kubeapps dashboard.

Reviewer: ```suggestion
At this point, the user having `EMAIL-ADDRESS` email account will have `cluster-admin` access and will be able to perform any desired action via the `kubectl` CLI or the Kubeapps dashboard.
```

### Step 2.5: Log in to Kubeapps with OIDC

Once Kubeapps is installed and configured, the next step is to log in and access the Kubeapps Web dashboard. The procedure to do this depends on how Kubeapps was configured.

1. If [the service was exposed externally](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps#exposing-externally), it may be accessed using a public IP address; if not, it can be accessed locally by forwarding the cluster port using the command below:

    ```bash
    kubectl port-forward -n kubeapps svc/kubeapps 8080:80
    ```

    This will start an HTTP proxy for secure access to the Kubeapps dashboard.

> 2. Browse to [http://127.0.0.1:8080](http://127.0.0.1:8080) (when forwarding the port) or to the public IP address of the serevice (when exposing the service externally). You see the Kubeapps login page, as shown below:

Reviewer: I might have missed it, but if they are using a public IP of the service, then the callback URI would need to be updated in the IdP (CSP) to include it?

Author: That's true, I should have mentioned it explicitly (I just mention _(this will be updated after Kubeapps is installed)_). I've added more details here. Thanks for pointing this out!
